### PR TITLE
feat: next

### PR DESCRIPTION
**Pull request by MarshallOfSound**
This is a proposed new release of `electron-download` being renamed to `@electron/download` and brought into the modern era 👍

### CODE REVIEW

File: src/downloader-resolver.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { Downloader } from './Downloader';
+import { GotDownloader } from './GotDownloader';
+
+export async function getDownloaderForSystem(): Promise<Downloader<any>> {
+  // TODO: Resolve the downloader or default to GotDownloader
+  // Current thoughts are a dot-file traversal for something like
+  // ".electron.downloader" which would be a text file with the name of the
+  // npm module to import() and use as the downloader
+  return new GotDownloader();
```
Comment (felixrieseberg) [inline src/downloader-resolver.ts]: Ah! Could you use `await import` here and prevent loading `Got` at all unless we really need to?

File: src/index.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  });
+}
+
+/**
+ * Downlaods an artifact from an Electron release and returns an absolute path
+ * to the downloaded file.
+ *
+ * @param artifactDetails The information required to download the artifact
+ */
+export async function downloadArtifcat(_artifactDetails: ElectronArtifactDetails): Promise<string> {
```
Comment (malept) [inline src/index.ts]: 😸 
```suggestion
export async function downloadArtifact(_artifactDetails: ElectronArtifactDetails): Promise<string> {
```

File: src/types.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+export interface ElectronDownloadRequestOptions {
+  force?: boolean;
+  unsafelyDisableChecksums?: boolean;
+  cacheRoot?: string;
+  downloadOptions?: DownloadOptions;
+  mirrorOptions?: MirrorOptions;
+  downloader?: Downloader<any>;
+}
+
+export type ElectronPlatformArtficatDetails = {
```
Comment (malept) [inline src/types.ts]: 😸 
```suggestion
export type ElectronPlatformArtifactDetails = {
```

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+  try {
+    result = await fn(directory);
+  } catch (err) {
+    await fs.remove(directory);
+    throw err;
+  }
+
+  await fs.remove(directory);
+  return result;
+}
```
Comment (malept) [inline src/utils.ts]: This vs. a dependency like `tmp-promise` 🤔

File: test/utils.spec.ts [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  describe('withTempDirectory()', () => {
+    it('should generate a new and empty directory', async () => {
+      await withTempDirectory(async dir => {
+        expect(await fs.pathExists(dir)).toEqual(true);
+        expect(await fs.readdir(dir)).toEqual([]);
+      });
+    });
+
+    it('should return the value the function returnrs', async () => {
```
Comment (malept) [inline test/utils.spec.ts]:
```suggestion
    it('should return the value the function returns', async () => {
```

File: test/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            expect(opts).toStrictEqual(downloadOpts);
+            await fs.writeFile(targetPath, 'file');
+          },
+        },
+        downloadOptions: downloadOpts,
+      });
+    });
+  });
+
+  describe('downloadArtifcat()', () => {
```
Comment (malept) [inline test/index.spec.ts]:
```suggestion
  describe('downloadArtifact()', () => {
```

File: src/downloader-resolver.ts [hunk 7]
```diff
@@ -0,0 +1,10 @@
+import { Downloader } from './Downloader';
+
+export async function getDownloaderForSystem(): Promise<Downloader<any>> {
+  // TODO: Resolve the downloader or default to GotDownloader
+  // Current thoughts are a dot-file traversal for something like
+  // ".electron.downloader" which would be a text file with the name of the
+  // npm module to import() and use as the downloader
```
Comment (malept) [inline src/downloader-resolver.ts]: Is the TODO meant to be resolved before the first release?

File: package.json [hunk 8]
```diff
@@ -0,0 +1,37 @@
+{
+  "name": "@electron/download",
+  "version": "1.0.0",
+  "description": "Utility for downloading different versions of Electron",
+  "main": "dist/cjs/index.js",
+  "module": "dist/esm/index.js",
```
Comment (malept) [inline package.json]: 🥂 Fancy.

File: .npmignore [hunk 9]
```diff
@@ -0,0 +1,4 @@
+*
+!package.json
+!README.md
+!dist/*
```
Comment (malept) [inline .npmignore]: Maybe set `files` in `package.json` instead?

File: .vscode/settings.json [hunk 10]
```diff
@@ -0,0 +1,5 @@
+{
+    "editor.formatOnSave": true,
+    "editor.tabSize": 2,
+    "typescript.preferences.quoteStyle": "single"
+}
```
Comment (malept) [inline .vscode/settings.json]: Unfortunate that this file and `.prettierrc.json` overlap.

I've started putting [EditorConfig](https://editorconfig.org/) files in some of my repos.

File: appveyor.yml [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+
+branches:
+  only:
+    - master
+
+environment:
+  matrix:
+  - nodejs_version: "4"
+  - nodejs_version: "6"
+  - nodejs_version: "8"
```
Comment (malept) [inline appveyor.yml]: Test against Node 6/8/10

File: src/utils.ts (hunk 4 above)
Comment (MarshallOfSound) [reply src/utils.ts]: The less deps the better IMO

File: src/downloader-resolver.ts (hunk 7 above)
Comment (MarshallOfSound) [reply src/downloader-resolver.ts]: Not required no, it's a Some Point In The Future TODO

File: .vscode/settings.json (hunk 10 above)
Comment (MarshallOfSound) [reply .vscode/settings.json]: Until editor config supports quote type we're going to have to duplicate

**Comment by MarshallOfSound** [comment]
> Missing .travis.yml? This git magic makes it confusing what's going to happen when the merge happens.

Once this is moved to `electron/download` I'll set up Circle CI

**Comment by malept** [comment]
Ugh, latest `got` requires Node 8. Should we follow suit now that April is fast approaching?

**Comment by MarshallOfSound** [comment]
> Ugh, latest got requires Node 8. Should we follow suit now that April is fast approaching?

Yeah, let's do it 👍

File: test/GotDownloader.spec.ts [hunk 12]
```diff
@@ -0,0 +1,75 @@
...
+        );
+        expect(await fs.pathExists(testFile)).toEqual(true);
+        expect(await fs.readFile(testFile, 'utf8')).toMatchSnapshot();
+      });
+    });
+
+    it('should throw an error if the file does not exist', async function() {
+      if (process.platform === 'win32') {
+        return;
+      }
```
[the remark this answers is not in the document]
Comment (MarshallOfSound) [reply test/GotDownloader.spec.ts]: 🤷‍♂️ It's a bug in `fs-extra`, once I have a windows machine I can look at it 👍

**Comment by MarshallOfSound** [comment]
> Are we dropping the progress bar?

Yes, if someone wants a progress bar they can implement a custom downloader 👍

**Comment by malept** [comment]
:shipit: Drop my request for changes and merge, I'm on mobile

**Pull request merged**